src/dmd/toir.d

```diff
@@ -904,7 +934,8 @@ void buildClosure(FuncDeclaration fd, IRState *irs)
         /* Build type for closure */
         type *Closstru = type_struct_class(closname, Target.ptrsize, 0, null, null, false, false, true, false);
         free(closname);
-        symbol_struct_addField(Closstru.Ttag, "__chain", Type_toCtype(Type.tvoidptr), 0);
+        auto chaintype = getParentClosureType(irs.sthis, fd);
+        symbol_struct_addField(Closstru.Ttag, "__chain", chaintype, 0);
```

Comment:
Future consideration, this field could be added only if required, and only if the outer frame context != this parameter.

It would save allocating an extra pointer size at run-time if the chain pointer is always null.

Reply:
Yes, but beware: a non-zero offset is also used to detect whether a variable is in a closure: https://github.com/dlang/dmd/blob/master/src/dmd/e2ir.d#L1175